### PR TITLE
Grandpa warp sync request-response protocol

### PR DESCRIPTION
**Pull request by expenses**
See #1208. A messy first draft of a full-node request-response protocol for use with https://github.com/paritytech/substrate-lite/pull/294. Essentially this takes @cheme's branch https://github.com/paritytech/substrate/compare/master...cheme:light_grandpa_warp and adds a request-response protocol on top.

### CODE REVIEW

File: client/network/src/config.rs
Comment (tomaka) [inline client/network/src/config.rs:110]: Considering that the actual protocol handling is not inside of `sc-network` but in a separate crate, I'd suggest to remove this field and instead ask the user to put the protocol in the `request_response_protocols` field, which is more generic. After all, nothing inside of `sc-network` distinguishes `NetworkConfiguration::grandpa_warp_sync_request_protocol_config` from `NetworkConfiguration::request_response_protocols`.

**Comment by tomaka** [comment]
Is this PR unfinished?
I'd have a few review comments, but is there a reason to not mark it as ready start the review process?

**Comment by expenses** [comment]
This PR is basically the minimum amount of changes I needed to make to @cheme 's branch in order to get grandpa warp sync to work in substrate-lite. There are definately some things that need to be changed before it can be merged, getting the inspect subcommand working again is a big one. I'll change it to allow reviews anyway though.

File: client/grandpa-warp-sync/Cargo.toml
Comment (tomaka) [inline client/grandpa-warp-sync/Cargo.toml:2]: I guess that will change before merging

File: client/grandpa-warp-sync/src/lib.rs
Comment (tomaka) [inline client/grandpa-warp-sync/src/lib.rs:65]:
```suggestion
		max_request_size: 32,
```

Matches what we did in the substrate-lite PR.

File: client/grandpa-warp-sync/src/lib.rs
Comment (tomaka) [inline client/grandpa-warp-sync/src/lib.rs:99]: Well, that comment isn't true for warp syncing.

File: client/grandpa-warp-sync/src/lib.rs
Comment (tomaka) [inline client/grandpa-warp-sync/src/lib.rs:68]: This timeout should probably be waaay lower. In practice, block requests take ~100-200ms, and a warp sync request should be much heavier. A 10 seconds timeout is IMO more than enough.

File: client/network/src/lib.rs
Comment (tomaka) [inline client/network/src/lib.rs:256]: Shouldn't be public. I believe that everything you need is re-exported from the `config` module (which is public).

File: client/finality-grandpa/src/finality_proof.rs
Comment (cheme) [inline client/finality-grandpa/src/finality_proof.rs:689]:
```suggestion
	// Fragment without change only allowed for proof last block.
```

**Review by cheme — COMMENTED** [review]
I see a few TODO from my branch, will try to remove them locally and also check if 'BlockJustification' trait is of any use (can be a remnant of something else that I did not clean).

File: client/grandpa-warp-sync/src/lib.rs
Comment (tomaka) [inline client/grandpa-warp-sync/src/lib.rs:1]:
```suggestion
// Copyright 2021 Parity Technologies (UK) Ltd.
```

**Review by tomaka — COMMENTED** [review]
Networking stuff looks good to me, but I can't review the rest

File: client/finality-grandpa/src/finality_proof.rs
Comment (cheme) [inline client/finality-grandpa/src/finality_proof.rs:760]: After checking this trait is only useful to keep the J parameter in the check method, this is similar to existing code, but since there is no use of TestFinality it is not strictly useful. I guess the right move is to create a test case that use prove and check, but at this point I did not check if implementing it for TestFinality would be straight forward.

**Comment by cheme** [comment]
@expenses , I just push some change to code from my branch, if you can check you agree with it.

File: client/grandpa-warp-sync/src/lib.rs
Comment (cheme) [inline client/grandpa-warp-sync/src/lib.rs:111]:
```suggestion
		let response = sc_finality_grandpa::prove_warp_sync(
```

File: client/grandpa-warp-sync/src/lib.rs
Comment (cheme) [inline client/grandpa-warp-sync/src/lib.rs:23]:
```suggestion
```

**Comment by expenses** [comment]
@cheme what do you think still needs to be done for this?

**Comment by cheme** [comment]
the One thing I am thinking about is really to forbid request that create huge proofs (IIUC the use case for light client is to warp sync from a snapshot, so limiting the number of block to the bounding duration is what I am thinking of but I may be wrong on the right security requirements).
I just realize that my comment about header cache is wrong (I was thinking we did only access the header that have a authority set change, but in fact we scan all headers so a specific cache should be use to avoid this scan, but can probably be done in a next step).

**Comment by cheme** [comment]
Thinking twice, if we want to warp from genesis as a first step, then the cache would be the must have (I don't remember if the snapshot was merged).

**Comment by tomaka** [comment]
> (IIUC the use case for light client is to warp sync from a snapshot, so limiting the number of block to the bounding duration is what I am thinking of but I may be wrong on the right security requirements).

The maximum size of the snapshot is always for the range `genesis .. best_block`. There's no way an attacker can maliciously request a proof that would be bigger than that.

But if it's too big, we could totally limit the number of blocks returned, so that clients would need to make multiple queries. That's not a big issue for the client.

**Comment by cheme** [comment]
Could either be limit allowed query 'start_block' param, return error when 'start_block' out of range.

Or limit size of produced proof: proof limited in number of authority set changes. 

I was thinking of the first one (reduce range for cache), but yeah the second one might be more flexible and easier to work with.

**Comment by expenses** [comment]
Do we want to add a cache before merging or can it wait?

**Comment by tomaka** [comment]
For what it's worth, the request-response system in the networking only allows processing one request (e.g. one GrandPa warp sync request) at a time. This means that even if the processing of a request took 10 seconds, the node would slow down but you can't abuse it to kill the node.

With that in mind, to me a cache isn't strictly needed. We're not opening a DoS vector.

On one hand, a cache isn't strictly needed to experiment with this. On the other hand, we all know that if we don't add a cache now, it will be left aside as "we'll do it later", and it might never be done.

**Comment by expenses** [comment]
> On the other hand, we all know that if we don't add a cache now, it will be left aside as "we'll do it later", and it might never be done.

IMO it'd be better to get this merged now if we can, but you're right. This is kind-of a company-wide problem though.

**Comment by cheme** [comment]
Added a test and a default limit for the proof.
While writing the test, I did realize that 'delay' is not handled correctly.
IIRC I am not sure 'delay' is currently use?
Also what happens when you define a set change and a new set change is define during the delay?
And what happens when a set change is pending and a forced set change is define during the delay?
And what happens when a forced set change is pending and a forced set change is define during the delay?
Will look in the code later, but if anybody got answer to these question in between, it can help :)  cc\ @andresilva

**Comment by andresilva** [comment]
> Also what happens when you define a set change and a new set change is define during the delay?
> And what happens when a set change is pending and a forced set change is define during the delay?
> And what happens when a forced set change is pending and a forced set change is define during the delay?

The delay should only be non-zero on forced changes. Additionally, there is a guarantee that changes never overlap each other, i.e. if a change was signalled at block #10 with a delay of 5, then you can be sure that no other pending change will exist between #10-#15. This only applies for the same fork, i.e. you might have a change at #11 on some other fork.

**Comment by cheme** [comment]
Thanks, good to hear, so the code is probably fine.
Should left only the issue of when the proof is queried for a starting block that is during a force fork delay.
The only solution I am thinking of is to always look backward for the previous authority set change, seems costy for a corner case, but at the same time it is not really recoverable (except running the query at a past starting block).

**Comment by cheme** [comment]
> i.e. you might have a change at #11 on some other fork.
@andresilva , just to confirm, in this case the change at #10 is ignored in favor of the fork at #11?

**Comment by cheme** [comment]
I did change a bit the iteration to not include change of authority set when another change is defined during the delay, should handle this case.

I also added a simple cache (last two commit), it can be remove to another pr if needed.

**Comment by tomaka** [comment]
Could you please merge master

**Comment by tomaka** [comment]
Could you please merge master so that the CI fixes itself

File: client/grandpa-warp-sync/Cargo.toml
Comment (andresilva) [inline client/grandpa-warp-sync/Cargo.toml:3]: Why create a new crate for this? I think it should just go in the existing grandpa crate. This also avoids having to expose some grandpa internals.

**Review by andresilva — CHANGES_REQUESTED** [review]
After #7339 we should be able to rewrite the finality proof changes much more clearly. The main change being that we will have a mapping of `set_id -> last_block_in_set`, which means that we can just fetch justifications from disk based on that rather than iterating through all headers to find them. I have reviewed the changes in `finality_proof` lightly since based on this I expect this code to be rewritten soon.

I don't want to block this PR any further on it (I can make those changes after #7339 is merged). My only gripe right now is that the warp sync code should exist inside the grandpa crate.

**Comment by andresilva** [comment]
> @andresilva , just to confirm, in this case the change at #10 is ignored in favor of the fork at #11?

I had not reviewed the PR yet when I answered this question. For our purposes, as long as we find a justification then we can be sure it is the correct one. Otherwise this would mean that GRANDPA finalized contradicting things (i.e. different forks). The existing code lgtm.

File: client/grandpa-warp-sync/Cargo.toml
Comment (expenses) [reply client/grandpa-warp-sync/Cargo.toml:3]: It's better to have it as a standalone crate IMO as it only can be used when BABE is also being used afaik.

Comment (andresilva) [reply client/grandpa-warp-sync/Cargo.toml:3]: As it exists now I think it's independent from the block production protocol, it's creating finality proofs from genesis to some latest finalized block regardless of how that block itself is validated (although it's true that we also need to fetch BABE data elsewhere). I won't press on moving this to the grandpa crate but can we rename it to `sc-finality-grandpa-warp-sync` so the naming is closer to the other crate?

**Comment by tomaka** [comment]
Merging master should hopefully now fix CI.

**Comment by tomaka** [comment]
I'd like to merge this today or tomorrow so that it is included in the upcoming Polkadot 0.7.28.
If someone has any last minute objection, please speak.

**Review by tomaka — APPROVED** [review]

**Review by andresilva — APPROVED** [review]

**Comment by andresilva** [comment]
Approving so we won't block this any further (as it blocks work on smoldot). But the `finality_proof.rs` changes will be refactored/rewritten after #7339 is in (I can create a PR for it afterwards).

**Comment by tomaka** [comment]
(I DM'ed @cheme and no objection either)

**Comment by tomaka** [comment]
bot merge

**Pull request merged**
